authorization/token/service/token_service.go

```diff
@@ -0,0 +1,326 @@
+		// we can re-evaluate its privileges to determine whether they have changed.
+		// If the privileges are unchanged, then reset the token status
+		if loadedToken.HasStatus(tokenPkg.TOKEN_STATUS_STALE) && resourceExistsInToken {
+			// Query for all of the token's privileges
+			privileges, err := s.Repositories().TokenRepository().ListPrivileges(ctx, tokenID)
+			if err != nil {
+				return nil, errors.NewInternalError(ctx, err)
+			}
+
+			scopesChanged := false
+			permissionsChanged := false
+
+			// If the number of permissions in the token is different to what is stored in the database, then
+			// something has definitely changed
+			if len(privileges) != len(*tokenClaims.Permissions) {
+				permissionsChanged = true
+			} else {
+				// Compare the scopes to those contained in the current token
+				for _, tokenPermission := range *tokenClaims.Permissions {
+					// Retrieve the up to date scopes for the resource
+					privilegeCache, err := s.Services().PrivilegeCacheService().CachedPrivileges(ctx, identity.ID, *tokenPermission.ResourceSetID)
+					if err != nil {
+						return nil, errors.NewInternalError(ctx, err)
+					}
+
+					// Compare the scopes of the resource with the scopes in the token
+					scopes := privilegeCache.ScopesAsArray()
+					if !s.scopesEquivalent(tokenPermission.Scopes, scopes) {
+						scopesChanged = true
+						break
+					}
+
+					resourceFound := false
+
+					// Also confirm that the resources in the token's permissions match those stored in the database
+					for _, priv := range privileges {
+						if priv.ResourceID == *tokenPermission.ResourceSetID {
+							resourceFound = true
+						}
+					}
+
+					if !resourceFound {
+						permissionsChanged = true
+						break
+					}
+				}
+			}
+
+			// If the scopes haven't changed, and the permission set is the same, and the specified resouce ID is
+			// already contained in the current token, then reset the token status to valid and return an empty string
+			if !scopesChanged && !permissionsChanged {
+				loadedToken.Status = 0
+				err = s.Repositories().TokenRepository().Save(ctx, loadedToken)
+				if err != nil {
+					return nil, errors.NewInternalError(ctx, err)
+				}
+
+				return nil, nil
+			}
+		}
+	}
+
+	// If we've gotten this far, it means that either no existing token was found, or the token that was found
+	// has been marked with status STALE and its privileges have changed, in either case we must generate a new token
+	signedToken := ""
+
+	err = s.ExecuteInTransaction(func() error {
+
+		// Initialize an array of permission objects that will be included in the token
+		perms := []token.Permissions{}
+
+		// Initialize an array of TokenPrivilege objects so that we can persist a record of the token's privileges to the database
+		tokenPrivs := []tokenRepo.TokenPrivilege{}
+
+		// Populate the scopes for the requested resource
+		privilegeCache, err := s.Services().PrivilegeCacheService().CachedPrivileges(ctx, identity.ID, resourceID)
```

Comment:
https://github.com/fabric8-services/fabric8-auth/pull/561/files#diff-1c4c6d8483d605cd1da70477e9a8442dR125

umm.. we would have fetched this data already in the previous block, but a re-computation wouldn't happen second time. So, wouldn't be expensive.

Reply:
No, the previous block fetches privileges for other permissions that already exist in the token. This particular privilege cache lookup is for the resource ID specified in the request.

Reply:
added the comment in the wrong place :)

I was talking about the overlap between 
```
			for _, tokenPermission := range *tokenClaims.Permissions {
					// Retrieve the up to date scopes for the resource
					privilegeCache, err := s.Services().PrivilegeCacheService().CachedPrivileges(ctx, identity.ID, *tokenPermission.ResourceSetID)
```
and

```
	for _, oldPriv := range oldTokenPrivs {
				// If we have hit the maximum permissions limit then break
				if len(perms) >= s.config.GetRPTTokenMaxPermissions() {
					break
				}

				// Don't process the same resource that was already specified for this request
				if oldPriv.ResourceID != resourceID {
					// Retrieve the cached privileges for the resource
					privilegeCache, err := s.Services().PrivilegeCacheService().CachedPrivileges(ctx, identity.ID, oldPriv.ResourceID)
```

Reply:
Right, so the second case might not even be hit in some circumstances, e.g. the token is flagged stale however after recalculation the scopes aren't the same.  Also, the second set of privilege cache lookups may not even encompass all of the privileges within the first test, in the situation where the token is at it's maximum limit for permissions.  We could potentially store the lookups in the first iteration in a local map variable and re-use that when populating the new token, however to be honest the db hit in this instance is going to be cheap as the privilege calculations have already been done.

Actually, what might be better here is if I introduce a many to one between the `TokenPrivilege` struct and `PrivilegeCache`, then preload the privilege cache records when calling `TokenRepository.ListPrivileges()`.  This will save repetitive db hits.  Leave this with me and I'll get it sorted out. :)

Reply:
I've fixed this now.

Reply:
Awesome!